File: common/data_plane/tests/test_qcrate_data_protocol.c
/* SPDX-License-Identifier: MIT */
#include "qcrate_data_protocol.h"

#include <stdio.h>
#include <string.h>

static int fail(const char *message)
{
	fprintf(stderr, "FAIL: %s\n", message);
	return 1;
}

int main(void)
{
	static const uint8_t expected_header[QCRATE_DATA_HEADER_BYTES] = {
		0x51, 0x43, 0x44, 0x50, 0x00, 0x01, 0x00, 0x40,
		0x00, 0x02, 0x00, 0x02, 0x00, 0x00, 0x00, 0x03,
		0x01, 0x02, 0x03, 0x04, 0x11, 0x12, 0x13, 0x14,
		0x15, 0x16, 0x17, 0x18, 0x21, 0x22, 0x23, 0x24,
		0x25, 0x26, 0x27, 0x28, 0x31, 0x32, 0x33, 0x34,
		0x35, 0x36, 0x37, 0x38, 0x41, 0x42, 0x43, 0x44,
		0x51, 0x52, 0x53, 0x54, 0x01, 0x23, 0x00, 0x04,
		0x61, 0x62, 0x63, 0x64, 0x65, 0x66, 0x67, 0x68,
	};
	struct qcrate_data_header header = {
		.packet_type = QCRATE_DATA_PACKET_DATA,
		.payload_format = QCRATE_DATA_FORMAT_IQ_S16_LE,
		.flags = QCRATE_DATA_FLAG_TIMESTAMP_VALID |
			 QCRATE_DATA_FLAG_FRAME_START,
		.stream_id = UINT32_C(0x01020304),
		.run_id = UINT64_C(0x1112131415161718),
		.shot_id = UINT64_C(0x2122232425262728),
		.packet_sequence = UINT64_C(0x3132333435363738),
		.frame_id = UINT32_C(0x41424344),
		.sample_offset = UINT32_C(0x51525354),
		.sample_count = UINT16_C(0x0123),
		.sample_bytes = 4,
		.timestamp_ticks = UINT64_C(0x6162636465666768),
	};
	struct qcrate_data_header decoded = {0};
	struct qcrate_data_stream_info info = {
		.frame_samples = 1024,
		.sample_rate_numerator = UINT64_C(12500000),
		.sample_rate_denominator = 1,
		.timestamp_rate_numerator = UINT64_C(200000000),
		.timestamp_rate_denominator = 1,
		.center_frequency_hz = -1000000,
		.channel_count = 2,
		.component_bits = 16,
		.fraction_bits = 15,
		.timestamp_clock_id = QCRATE_DATA_CLOCK_QCRATE_TIMEBASE,
		.config_id = UINT64_C(0x1020304050607080),
	};
	struct qcrate_data_stream_info decoded_info = {0};
	struct qcrate_data_shot_end end = {
		.status = QCRATE_DATA_SHOT_COMPLETE,
		.end_flags = QCRATE_DATA_FLAG_END_OF_STREAM,
		.total_frames = 4,
		.total_samples = 4096,
		.total_data_packets = 52,
		.first_packet_sequence = 2,
		.last_packet_sequence = 53,
		.payload_bytes = 16384,
		.payload_crc32 = UINT32_C(0x12345678),
	};
	struct qcrate_data_shot_end decoded_end = {0};
	uint8_t encoded[QCRATE_DATA_HEADER_BYTES];
	uint8_t object[QCRATE_DATA_STREAM_INFO_BYTES];

	if (qcrate_data_header_encode(encoded, &header) != QCRATE_DATA_CODEC_OK)
		return fail("header encode rejected a valid object");
	if (memcmp(encoded, expected_header, sizeof(encoded)) != 0)
		return fail("header does not match the cross-language golden vector");
	if (qcrate_data_header_decode(&decoded, encoded, sizeof(encoded)) !=
	    QCRATE_DATA_CODEC_OK)
		return fail("header decode rejected the golden vector");
	if (decoded.packet_sequence != header.packet_sequence ||
	    decoded.timestamp_ticks != header.timestamp_ticks ||
	    decoded.header_bytes != QCRATE_DATA_HEADER_BYTES)
		return fail("header round trip changed a field");

	if (qcrate_data_stream_info_encode(object, &info) != QCRATE_DATA_CODEC_OK ||
	    qcrate_data_stream_info_decode(&decoded_info, object, sizeof(object)) !=
	    QCRATE_DATA_CODEC_OK)
		return fail("stream-info round trip failed");
	if (decoded_info.center_frequency_hz != info.center_frequency_hz ||
	    decoded_info.sample_rate_numerator != info.sample_rate_numerator ||
	    decoded_info.config_id != info.config_id)
		return fail("stream-info round trip changed a field");

	if (qcrate_data_shot_end_encode(object, &end) != QCRATE_DATA_CODEC_OK ||
	    qcrate_data_shot_end_decode(&decoded_end, object, sizeof(object)) !=
	    QCRATE_DATA_CODEC_OK)
		return fail("shot-end round trip failed");
	if (decoded_end.total_samples != end.total_samples ||
	    decoded_end.payload_crc32 != end.payload_crc32)
		return fail("shot-end round trip changed a field");

	encoded[0] = 0;
	if (qcrate_data_header_decode(&decoded, encoded, sizeof(encoded)) !=
	    QCRATE_DATA_CODEC_BAD_MAGIC)
		return fail("bad magic was not rejected");

	puts("PASS: Q-Crate Data Plane v1 C codec");
	return 0;
}

File: host/data_plane/tests/test_qcrate_data_protocol.py
#!/usr/bin/env python3
"""Exact and malformed-vector tests for Q-Crate Data Plane v1."""
from __future__ import annotations

import sys
import unittest
from pathlib import Path


DATA_PLANE_DIR = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(DATA_PLANE_DIR))

import qcrate_data_protocol as protocol  # noqa: E402


GOLDEN_HEADER = bytes.fromhex(
    "51434450000100400002000200000003"
    "01020304111213141516171821222324"
    "25262728313233343536373841424344"
    "51525354012300046162636465666768"
)


def golden_header(**changes: object) -> protocol.Header:
    values: dict[str, object] = {
        "packet_type": protocol.PacketType.DATA,
        "payload_format": protocol.PayloadFormat.IQ_S16_LE,
        "flags": protocol.Flags.TIMESTAMP_VALID | protocol.Flags.FRAME_START,
        "stream_id": 0x01020304,
        "run_id": 0x1112131415161718,
        "shot_id": 0x2122232425262728,
        "packet_sequence": 0x3132333435363738,
        "frame_id": 0x41424344,
        "sample_offset": 0x51525354,
        "sample_count": 0x0123,
        "sample_bytes": 4,
        "timestamp_ticks": 0x6162636465666768,
    }
    values.update(changes)
    return protocol.Header(**values)  # type: ignore[arg-type]


class HeaderTests(unittest.TestCase):
    def test_matches_cross_language_golden_vector(self) -> None:
        header = golden_header()
        self.assertEqual(header.encode(), GOLDEN_HEADER)
        self.assertEqual(protocol.Header.decode(GOLDEN_HEADER), header)

    def test_rejects_magic_version_and_short_header(self) -> None:
        with self.assertRaisesRegex(ValueError, "shorter"):
            protocol.Header.decode(GOLDEN_HEADER[:-1])
        bad_magic = bytearray(GOLDEN_HEADER)
        bad_magic[0] = 0
        with self.assertRaisesRegex(ValueError, "magic"):
            protocol.Header.decode(bad_magic)
        bad_version = bytearray(GOLDEN_HEADER)
        bad_version[5] = 2
        with self.assertRaisesRegex(ValueError, "version"):
            protocol.Header.decode(bad_version)


class ObjectTests(unittest.TestCase):
    def test_stream_info_round_trip(self) -> None:
        info = protocol.StreamInfo(
            frame_samples=1024,
            sample_rate_numerator=12_500_000,
            sample_rate_denominator=1,
            timestamp_rate_numerator=200_000_000,
            timestamp_rate_denominator=1,
            center_frequency_hz=-1_000_000,
            channel_count=2,
            component_bits=16,
            fraction_bits=15,
            timestamp_clock_id=protocol.TimestampClock.QCRATE_TIMEBASE,
            config_id=0x1020304050607080,
        )
        self.assertEqual(protocol.StreamInfo.decode(info.encode()), info)

    def test_shot_end_round_trip_and_reserved_check(self) -> None:
        end = protocol.ShotEnd(
            status=int(protocol.ShotStatus.COMPLETE),
            end_flags=int(protocol.Flags.END_OF_STREAM),
            total_frames=4,
            total_samples=4096,
            total_data_packets=52,
            first_packet_sequence=2,
            last_packet_sequence=53,
            payload_bytes=16384,
            payload_crc32=0x12345678,
        )
        encoded = end.encode()
        self.assertEqual(protocol.ShotEnd.decode(encoded), end)
        malformed = bytearray(encoded)
        malformed[15] = 1
        with self.assertRaisesRegex(ValueError, "reserved"):
            protocol.ShotEnd.decode(malformed)


class PacketTests(unittest.TestCase):
    def test_stream_info_packet_round_trip(self) -> None:
        info = protocol.StreamInfo(
            frame_samples=1024,
            sample_rate_numerator=12_500_000,
            sample_rate_denominator=1,
            timestamp_rate_numerator=200_000_000,
            timestamp_rate_denominator=1,
            center_frequency_hz=1_000_000,
            channel_count=2,
            component_bits=16,
            fraction_bits=15,
            timestamp_clock_id=protocol.TimestampClock.QCRATE_TIMEBASE,
            config_id=1,
        )
        packet = protocol.Packet(
            protocol.Header(
                packet_type=protocol.PacketType.STREAM_INFO,
                payload_format=protocol.PayloadFormat.IQ_S16_LE,
                flags=protocol.Flags(0),
                stream_id=1,
                run_id=2,
                shot_id=0,
                packet_sequence=0,
                sample_bytes=4,
            ),
            info.encode(),
        )
        self.assertEqual(protocol.Packet.decode(packet.encode()), packet)

    def test_data_packet_round_trip(self) -> None:
        payload = bytes.fromhex("0100010002000200")
        header = golden_header(
            flags=protocol.Flags.FRAME_START,
            frame_id=0,
            sample_offset=0,
            sample_count=2,
            timestamp_ticks=0,
        )
        packet = protocol.Packet(header, payload)
        self.assertEqual(protocol.Packet.decode(packet.encode()), packet)

    def test_rejects_truncated_wrong_format_and_oversize_data(self) -> None:
        header = golden_header(
            flags=protocol.Flags.FRAME_START,
            sample_offset=0,
            sample_count=2,
            timestamp_ticks=0,
        )
        with self.assertRaisesRegex(ValueError, "expected 8"):
            protocol.Packet(header, b"1234").encode()
        with self.assertRaisesRegex(ValueError, "requires 4"):
            protocol.Packet(
                golden_header(sample_bytes=2), b"1234"
            ).encode()
        with self.assertRaisesRegex(ValueError, "exceeds"):
            protocol.Packet(
                golden_header(
                    flags=protocol.Flags.FRAME_START,
                    sample_offset=0,
                    sample_count=335,
                    timestamp_ticks=0,
                ),
                bytes(335 * 4),
            ).encode()

    def test_rejects_false_timestamp_and_bad_frame_start(self) -> None:
        with self.assertRaisesRegex(ValueError, "TIMESTAMP_VALID"):
            protocol.Packet(
                golden_header(flags=protocol.Flags(0)), bytes(0x0123 * 4)
            ).encode()
        with self.assertRaisesRegex(ValueError, "sample offset"):
            protocol.Packet(
                golden_header(flags=protocol.Flags.FRAME_START, timestamp_ticks=0),
                bytes(0x0123 * 4),
            ).encode()

    def test_rejects_metadata_sample_fields_and_format(self) -> None:
        base = {
            "packet_type": protocol.PacketType.HEARTBEAT,
            "payload_format": protocol.PayloadFormat.NONE,
            "flags": protocol.Flags(0),
            "stream_id": 1,
            "run_id": 2,
            "shot_id": 3,
            "packet_sequence": 4,
        }
        with self.assertRaisesRegex(ValueError, "NONE requires 0"):
            protocol.Packet(protocol.Header(**base, sample_bytes=4)).encode()
        with self.assertRaisesRegex(ValueError, "must not identify"):
            protocol.Packet(protocol.Header(**base, frame_id=1)).encode()


if __name__ == "__main__":
    unittest.main()
